library/kani/src/lib.rs

```diff
@@ -65,6 +65,29 @@ pub fn assume(cond: bool) {
     assert!(cond, "`kani::assume` should always hold");
 }
 
+/// If the `premise` is true, so must be the `conclusion`
+///
+/// Note that boolean operators (such as `||`) are evaluated lazily by Rust.
+/// This function is not and both conditions will be evaluated always. As a
+/// reult this function is not intended to be used in regular code. Instead it
```

Comment:
```suggestion
/// result this function is not intended to be used in regular code. Instead it
```

Reply:
Wouldn't the performance be worse for verification too?

Reply:
I don't think so? I don't think the verification performs any sort of short circuiting of boolean expressions. Or at least not in such a way that using control flow would make a difference. I'm not sure though. I can remove it if you like.

Reply:
It does. Here is a dummy example of using eager imply:
```rust
pub fn implies(premise: bool, conclusion: bool) -> bool {
    !premise || conclusion
}

#[kani::proof]
fn check_eager() {
    let arg1 = false;
    let arg2 = || {
        let mut heads = false;
        loop {
            heads = !heads;
            if kani::any() {
                break
            }
        }
        heads
    };
    assert!(implies(arg1, arg2()));
}
```

and the same example but using a lazy implies:
```rust
pub fn lazy_implies<F>(premise: bool, mut conclusion: F) -> bool where F: FnMut() -> bool {
    !premise || conclusion()
}


#[kani::proof]
fn check_lazy() {
    let arg1 = false;
    let arg2 = || {
        let mut heads = false;
        loop {
            heads = !heads;
            if kani::any() {
                break
            }
        }
        heads
    };
    assert!(lazy_implies(arg1, arg2));
}
```
The first harness won't terminate while the second verifies in less than a second.

Reply:
That surprises me. What should we do about it? I could remove `implies` and force people to use the expanded version instead. I can also add syntactic sugar for it that expands it out.

Reply:
I would remove it for now and create a separate PR with a more thorough performance analysis to see which one performed better in different scenarios